### DISCUSSION TITLE
Louvain algorithm seed parameter doesn't work

### DISCUSSION THREAD

**Discussion by momertaylan** [networkx/networkx#6129] [Q&A]
I am trying to use community detection algorithm in my project. I am using the louvain algorithm with seed parameters but whenever I rerun the code, the result of louvain algorithm gives different results.

When I run the code from documentation example, the louvain algorithm with seed parameters works very well:
import networkx as nx
import networkx.algorithms.community as nx_comm
G = nx.petersen_graph()
nx_comm.louvain_communities(G, seed=123)

Then I created my graph by using nx.from_pandas_edgelist:
g1= nx.from_pandas_edgelist(df,"Customer1","Customer2",edge_attr=["Relation"],
                        create_using= nx.MultiGraph())

After that I run louvain algorithm 2 times with the same seed parameter but it gives different result.

Why seed parameter doesn't work in this situation?

Number of Nodes: 3.5 million
Number of Edges:  7.5 million

**Comment by dschult** [comment]
Can you show the code you use to run the communities on the larger graph?  I mean, you put the code to read the edgelist, but not the code that actually causes the problem.  Can you try a very small example, say with 20 nodes and see if you have the same issue?

**Reply by momertaylan** [reply]
I run the same code as documentation example after creating graph from edgelist: 

import networkx as nx
import networkx.algorithms.community as nx_comm
g1= nx.from_pandas_edgelist(df,"Customer1","Customer2",edge_attr=["Relation"],create_using= nx.MultiGraph())
communities_louvain = nx_comm.louvain_communities(g1, seed=123,resolution=10)

I tried on small example and I don't get any issue. I get same result from two different runs with same seed.

Can this be a problem on larger graph?

**Reply by dschult** [reply]
I can't think of a reason it would change for a large graph.  Maybe round-off error? I don't know. But it is easy to try a large graph:
```python
import networkx as nx
import networkx.algorithms.community as nx_comm
g1=nx.binomial_graph(100000, 0.3, seed=123)
c1 = nx_comm.louvain_communities(g1, seed=123, resolution=10)
c2 = nx_comm.louvain_communities(g1, seed=123, resolution=10)
set(c1)==set(c2)
```
How are you checking that the result is different between different runs with the same seed?

**Reply by momertaylan** [reply]
I assign an id to each community sets and then get community id and node number into a dataframe:

```python
import networkx as nx
import networkx.algorithms.community as nx_comm
communities_louvain = nx_comm.louvain_communities(G1, seed=123,resolution=10)
#creating community id
dct={}
for ind,item in enumerate(communities_louvain):
    dct[ind+1]=item
df_com = pd.DataFrame.from_dict(dct, orient='index').T.melt(var_name='CommunityID', value_name='IP_ID').dropna(subset=['IP_ID'])
```

Then I start kernel again and run the same code above. 

After getting 2 dataframe I check manually whether 2 node number is in same community or not.

**Reply by rossbar** [reply]
What happens when you run @dschult's code snippet, replacing the `g1` from his example with the `G1` from yours?

**Comment by momertaylan** [comment]
Can this be a problem on Multigraph? @dschult @rossbar

**Reply by rossbar** [reply]
It's going to be difficult to say anything without a minimal reproducing example. Did you try @dschult's snippet? Note in @dschult 's example that the *sets* are compared - i.e. the ordering of the nodes within the communities are ignored. In your conversion to DataFrames, this is explicitly no longer the case, so you could get "wrong" answers with the same seed due to different orderings.

**Reply by momertaylan** [reply]
After converting to DataFrame, I check every communities with each other by using _sets_ like @dschult 's snippet but it doesn't work.

**Reply by rossbar** [reply]
As mentioned above, it's going to be difficult to say anything concrete without a minimal reproducing example. A good minimal reproducing example has the following properties:
 - It is as small/narrowly focused as possible. For example, if you can limit the components that involve pandas and focus only on the NX components, that will help.
 - it doesn't involve a ton of data. What's the smallest possible graph you can reproduce this on? Even the exercise of trying to identify the smallest possible reproducing input might make it clear what the problem is.

**Reply by momertaylan** [reply]
I can not directly use NX component because I need to test the result with production server and my server with same data and same NX version. This is why I need to convert all communities to pandas and then try to compare these two results. 

Do you have any suggestion to test my community result with production result?

**Reply by dschult** [reply]
I think what we're saying is that we can't replicate your problem. We can't show that anything is wrong. To debug your code you will need to separate the different parts. As far as we can tell the NetworkX side of the process is working fine. To help us help you, you will need to separate the other components of your code to provide an example of networkx not working correctly. We can't debug something if it isn't happening in our code. Try to find an example (perhaps part of your dataset) that is simple enough to allow us to run it and which still produces this error.  Often in the process of separating the various parts of the code, you will find the error.
